### PR TITLE
feat: Research updates upload and download of files

### LINKED ISSUE #2238: [feature request] Add file download to research
**Is your feature request related to a problem? Please describe.**
We can't add files to research topics to download. (besides an external link)

**Describe the solution you'd like**
Would be nice to be able to upload files to a research, like we made in how-to's. https://github.com/ONEARMY/community-platform/pull/1664
This keeps track of the downloads and makes sure the file stays online. The button can be added per step

**Describe alternatives you've considered**
Continue using Gdrive to store files.

**What it would look like**

download

Features:

- [ ] Add the option to upload files/links in a research update
- [ ] Add the counter for downloads for files in updates 
- [ ] Refactor and implement firebase incrementation functions

### PR DESCRIPTION
**Pull request by balhelhas**
PR Checklist

- [ x ] - Commit [messages are descriptive](https://github.com/ONEARMY/community-platform/blob/master/CONTRIBUTING.md#--commit-style-guide), it will be used in our [Release Notes](https://github.com/ONEARMY/community-platform/releases/)

PR Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Developer experience (improves developer workflows for contributing to the project)

## Description

This PR includes the upload and download feature in the Research updates. It was been done extremely equal to the same feature in the How To's with a change into the download count, which it seemed overly complex. In my approach I added a variable downloadCount into each update and incremented it once someone downloads a file. 

I also made a change to the upload button, don't know if it makes sense. The label was Upload files (.zip) but there is no validation on the upload it is possible to upload other types of files, plus maybe for the research module it isn't the best to be strict to only zip files.
If we must keep this as restriction we should create a validation when uploading. 

## Git Issues

Closes #2238

## Screenshots/Videos

---

## What happens next?

Thanks for the contribution! We try to make sure all PRs are reviewed ahead of a monthly dev call (first Monday of the month, open to all!).

If the PR is working as intended it'll be merged and included in the next platform release, if not changes will be requested and re-reviewed once updated.

If you need more immediate feedback you can try reaching out on Discord in the [Community Platform `development` channel](https://discord.com/channels/586676777334865928/938781727017558018).

### CODE REVIEW

**Comment by balhelhas** [comment]
I haven't added tests for it yet. Adding them  ASAP

**Comment by davehakkens** [comment]
how do you test this @balhelhas. You assign the `research_creator` role to yourself?

File: src/pages/Research/Content/Common/ResearchUpdate.form.tsx [hunk 1]
```diff
@@ -457,17 +508,21 @@ const getResearchUpdates = (
...
+    if (images.length) return 'Do not include both images and video'
+
     const ytRegex = new RegExp(/(youtu\.be\/|youtube\.com\/watch\?v=)/gi)
     const urlValid = ytRegex.test(videoUrl)
     return urlValid ? null : 'Please provide a valid YouTube Url'
   }
-  return images && images[0] ? null : 'Include either images or a video'
+
+  if (!images.length && !files.length && !videoUrl)
+    return 'Include either files, images or video'
```
Comment (evakill) [inline src/pages/Research/Content/Common/ResearchUpdate.form.tsx:527]: wanted to double check on this req @davehakkens - if the user includes a download in the research, do they still have to include image or video?

**Review by evakill — APPROVED** [review]
code looks great! before merging:
- make sure @davehakkens can do design review 
- add a test (check out the [howto tests](https://github.com/ONEARMY/community-platform/pull/2716/commits/54ec009a6119b729e7abb16d61b02f84049b0159#diff-e9c6e3e01cb8e1e56c516669dfaff52bcb7015460a2d2a6cc701e2d107995964R182-R208) for a good example)
- fix the failing unit test (not sure why that one is failing, if you're stuck we can try to debug together next week)

**Comment by evakill** [comment]
> how do you test this @balhelhas. You assign the `research_creator` role to yourself?

i was actually able to test this one with just admin and beta-tester roles - are you not able to? also, FYI i realized that demo_admin@example.com only works when running locally, not on the preview site

**Comment by davehakkens** [comment]
Got it!
Looks good but two remarks:
1. Can we put it at the bottom of the form. So the images/video upload comes first?
2. Can we also add the option to add external files like we have in the how-to?  (I conisdered this as part of the download feature, but realise now its not that obvious. Sorry for that!)
afbeelding

File: src/pages/Research/Content/Common/ResearchUpdate.form.tsx (hunk 1 above)
Comment (davehakkens) [reply src/pages/Research/Content/Common/ResearchUpdate.form.tsx:527]: well spotted @evakill . Ideally it stays the same as before. Adding an image or video is mandatory. Adding files is optional.
(we're still discussing if we should change this but thats for another issue)

**Comment by balhelhas** [comment]
Thanks for the feedback I will work this improvements.

**Comment by balhelhas** [comment]
@davehakkens @evakill when you have the time please double check it. I made the changes mentioned above, still missing the test. Once that is done and you have confirmed everything is OK we could merge it, thanks!

**Comment by davehakkens** [comment]
nice thanks for the quick changes!
Could w change the wording from
`Set your file(s) for this update` to `Attach your file(s) for this update`


And when I add the 'download link' it seem to get stuck here. Does it work for you?
afbeelding

**Comment by balhelhas** [comment]
@davehakkens found the issue it should be working now

**Comment by davehakkens** [comment]
Nice. yep all good now from my side! @evakill final approval?

**Review by evakill — APPROVED** [review]
> add a test (check out the [howto tests](https://github.com/ONEARMY/community-platform/pull/2716/commits/54ec009a6119b729e7abb16d61b02f84049b0159#diff-e9c6e3e01cb8e1e56c516669dfaff52bcb7015460a2d2a6cc701e2d107995964R182-R208) for a good example)

LGTM - still want to get that unit test in, but feel free to merge after doing so!

**Comment by balhelhas** [comment]
Cool, I am working on them. Hope to finalize them today

**Pull request merged**